src/Compilers/CSharp/Portable/Syntax/NonRecursiveSyntaxWalker.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Represents a non-recursive visitor which descends an entire <see cref="CSharpSyntaxNode"/> graph
+    /// </summary>
+    public abstract partial class NonRecursiveSyntaxWalker : CSharpSyntaxVisitor<NonRecursiveSyntaxWalker.Chunk>
+    {
+        public class Chunk
```

Comment:
Actually I used a class for consistency with the rewriter but I can definitively use a struct here.